### ISSUE TITLE
2 questions about 3.0 release

### ISSUE THREAD

**Issue by ZXfkSIE** [complexlogic/loudgain#7]
1, Why have the strip tag option `-S` been removed?
2. Are you going to upload this project to software repositories of those popular Linux distros?

**Comment by complexlogic** [comment]
> Why have the strip tag option -S been removed?

I rewrote the `tag.cpp` file to use templates for the file formats, which significantly reduced the amount of duplicated code. The problem with the strip tags feature in TagLib is that it is not implemented consistently across the formats in the library, which creates problems for using templates. There were also a lot of bugs in TagLib related to the strip feature in the past. I wanted to avoid the complexity and make the code simple, easier to maintain, and have fewer bugs. 

Also, I think the feature is not that useful in 2022. ID3v2 has been the standard in MP3 files for more than 20 years now. And if people do still have ID3v1 tags, there are plenty of other tools that will remove them.

> Are you going to upload this project to software repositories of those popular Linux distros?

Typically, a maintainer of a distro has to submit the package and take responsibility for it. They don't accept submissions from the public. Currently, rsgain is available in the Arch Linux AUR repo only.

**Comment by ZXfkSIE** [comment]
With regard to the popular software repositories. They do accept submissions from the public.
- Fedora Copr (maintained by users completely) and RPM Fusion (need reviewing, see https://rpmfusion.org/Contributors) for Fedora and CentOS
- Launchpad PPA for Ubuntu
- For Debian it is possible to publish your software through a sponsor. See https://www.debian.org/doc/manuals/distribute-deb/distribute-deb.html#adding-packages-to-debian . After that it will also become available in Ubuntu.
- Other 3rd-party software package hosting services, such as https://packagecloud.io/

**Comment by ZXfkSIE** [comment]
Also, please consider making a snap or flatpak release. (I strongly recommend flatpak since it is supported by SteamOS)

**Comment by complexlogic** [comment]
@ZXfkSIE That is a good idea, I will look into flatpak.

**Comment by ZXfkSIE** [comment]
On Fedora 37:
```
 Problem: package rsgain-3.2-1.x86_64 requires libswresample-free >= 5, but none of the providers can be installed
  - package ffmpeg-libs-5.1.2-3.fc37.x86_64 conflicts with libswresample-free provided by libswresample-free-5.1.2-1.fc37.i686
  - package ffmpeg-libs-5.1.2-3.fc37.x86_64 conflicts with libswresample-free provided by libswresample-free-5.1.2-1.fc37.x86_64
  - package ffmpeg-libs-5.1.2-3.fc37.x86_64 conflicts with libswresample-free provided by libswresample-free-5.1.2-6.fc37.i686
  - package ffmpeg-libs-5.1.2-3.fc37.x86_64 conflicts with libswresample-free provided by libswresample-free-5.1.2-6.fc37.x86_64
  - problem with installed package ffmpeg-libs-5.1.2-9.fc37.x86_64
  - package ffmpeg-libs-5.1.2-9.fc37.x86_64 conflicts with libswresample-free provided by libswresample-free-5.1.2-1.fc37.i686
  - package ffmpeg-libs-5.1.2-9.fc37.x86_64 conflicts with libswresample-free provided by libswresample-free-5.1.2-1.fc37.x86_64
  - package ffmpeg-libs-5.1.2-9.fc37.x86_64 conflicts with libswresample-free provided by libswresample-free-5.1.2-6.fc37.i686
  - package ffmpeg-libs-5.1.2-9.fc37.x86_64 conflicts with libswresample-free provided by libswresample-free-5.1.2-6.fc37.x86_64
  - conflicting requests
```
That's why I suggest a cross-platform solution like flatpak. Solving dependency bugs and outdated problems (e.g. Debian distros still stick on taglib 1.11 while 1.13 is out now) on EVERY version of EVERY distro will cost your whole life LOL

**Comment by phw** [comment]
> @ZXfkSIE That is a good idea, I will look into flatpak.

While I personally much prefer Flatpak over Snap for various reasons, I think for this specific case Snap would be the better choice. Flatpak is really made for desktop applications and does not work that well for CLI tools. You can package them of course, but the need to be run with `flatpak run [app-fully-qualified-name]`. Users can of course place a custom script to run this, but that's not as convenient.

Snap would allow to run the app normally as `rsgain`. Just make sure to use classic confinement so it has access to files on the entire system.

**Comment by ZXfkSIE** [comment]
> > @ZXfkSIE That is a good idea, I will look into flatpak.
> 
> While I personally much prefer Flatpak over Snap for various reasons, I think for this specific case Snap would be the better choice. Flatpak is really made for desktop applications and does not work that well for CLI tools. You can package them of course, but the need to be run with `flatpak run [app-fully-qualified-name]`. Users can of course place a custom script to run this, but that's not as convenient.
> 
> Snap would allow to run the app normally as `rsgain`. Just make sure to use classic confinement so it has access to files on the entire system.

You are right. I saw Flathub required the author to submit an 128x128 icon, then the author gave up submitting this project to Flathub on yesterday LOL

**Comment by complexlogic** [comment]
Yes that is correct. I tried to make a flatpak but realized that it's not well-suited for CLI applications and withdrew the attempt. I will look into snap.

**Comment by atomGit** [comment]
i don't know what's involved, but perhaps an alternative to Snap/Flatpak if you wanted to avoid those (especially Snap since it's proprietary i believe) might be a static binary that includes dependencies? i imagine this would also ensure it works no matter where it's run

**Comment by ZXfkSIE** [comment]
> i don't know what's involved, but perhaps an alternative to Snap/Flatpak if you wanted to avoid those (especially Snap since it's proprietary i believe) might be a static binary that includes dependencies? i imagine this would also ensure it works no matter where it's run

In the past I compiled a C++ program under Fedora and then copied it into CentOS. It does not work.

**Comment by ZXfkSIE** [comment]
Another potential solution is use a unified package manager like Conan. Then people can download and compile Taglib 1.13 statically on EVERY Linux distros.

**Comment by Calinou** [comment]
If you want to produce a portable binary for CLI applications that you can distribute from anywhere, AppImage or Exodus may fit the bill. These don't have the concerns that Snap has (such as overhead, centralization, etc).

**Comment by AndrewKvalheim** [comment]
Related: NixOS/nixpkgs#220842

**Comment by complexlogic** [comment]
A few updates to share on the packaging situation.

rsgain is now an official Debian package. It is available in Debian 13 Trixie (current Testing), and will be available in Ubuntu 24.04.

I updated the build script to support static builds for Linux. I wasn't satisfied with any of the cross-distribution packaging options (flatpak, snap, etc), and decided static builds would be the best solution. This Linux static build should run on any GNU-based Linux distro released in the last 2 years.

Binaries for macOS will now be provided, with separate builds for x86 and ARM based macs. This will save mac users the hassle of having to set up Xcode and Homebrew, which may be confusing for non-developers.

With all of the above installation options now available, I believe that the packaging for this program is good enough. I'm going to close this issue.

**Comment by ZXfkSIE** [comment]
Sorry for reopening. Could you please add the instructions about how to statically build rsgain at https://github.com/complexlogic/rsgain/blob/master/docs/BUILDING.md ?
